src/main/java/com/google/devtools/build/lib/remote/util/NetworkTime.java

```diff
@@ -0,0 +1,123 @@
+// Copyright 2019 The Bazel Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package com.google.devtools.build.lib.remote.util;
+
+import static java.util.concurrent.TimeUnit.NANOSECONDS;
+
+import com.google.devtools.build.lib.concurrent.ThreadSafety;
+import com.google.common.base.Stopwatch;
+import io.grpc.CallOptions;
+import io.grpc.Channel;
+import io.grpc.ClientCall;
+import io.grpc.ClientInterceptor;
+import io.grpc.Context;
+import io.grpc.ForwardingClientCall;
+import io.grpc.ForwardingClientCallListener;
+import io.grpc.Metadata;
+import io.grpc.MethodDescriptor;
+import io.grpc.Status;
+import java.time.Duration;
+
+/** Reentrant wall clock stopwatch and grpc interceptor for network waits. */
```

Comment:
Why do you need the reentrant part? I don't follow why this interceptor should be more complex than take time on first call and stop taking time on close?

Reply:
Because uploads and downloads occur concurrently and asynchronously during ensureInputsProvided and remoteCache.download respectively, and have start and stop times interspersed throughout their blocking periods of operations, with possible times during which no network activity is outstanding

Reply:
I think internally, Google is computing network time as total wall time minus all known sub-intervals. Maybe that's good enough here as well?